### ISSUE TITLE
Optimiser Layer

### ISSUE THREAD

**Issue by karanchahal** [karanchahal/buildTensorflow#17]
This issue tracks the development of the optimiser layer. Optimisers like RMSProp, SGD will extend from this.

**Comment by karanchahal** [comment]
In the process of implementing this right now. The functionality we'd like to support is the updating of our tensors with their respective gradients using the Optimiser flavour we want (SGD, RMSProp, Adam etc).

Hence there would be a base optimiser class  with a variable called params, which has a list of all tensors that are in the computational graph and the ones that need to be updated. 

```c++
template<typename T>
class Optimizer {
    
    public:
    vector<Tensor<T>*> params;

    void zeroGrad() {
        for(auto i : params) {
            i->zeroGrad();
        }
    }

    virtual void step(T learning_rate) {};
};

```
Some noteworthy points:

1. Here we have a function step that will be overridden according to optimiser.
2. We need a function zeroGrad to rests all the gradients of  the tensors for the next forward prop and backward Prop cycle.

**Comment by karanchahal** [comment]
Handled by #22
